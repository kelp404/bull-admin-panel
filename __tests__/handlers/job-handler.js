const util = require('util');
const config = require('config');
const Bull = require('bull');
const redis = require('redis');
const Request = require('../../lib/models/request');
const Response = require('../../lib/models/response');
const jobHandler = require('../../lib/handlers/job-handler');
let queues;

beforeEach(() => {
  const client = redis.createClient(config.bull.redis);
  const flushDB = util.promisify(client.flushdb).bind(client);

  return flushDB()
    .then(() => client.end(true))
    .then(() => {
      queues = [new Bull('test', config.bull)];
    });
});

afterEach(() => {
  jest.restoreAllMocks();
  queues.forEach(queue => queue.close());
});

const generateRequest = ({method, url}) => {
  return new Request({
    id: Math.random().toString(36).substr(2),
    queues,
    method,
    url
  });
};

const generateResponse = requestId => {
  return new Response({
    requestId,
    ws: {}
  });
};

describe('get job', () => {
  test('get a waiting job by id', () => {
    const request = generateRequest({
      method: 'GET',
      url: '/queues/test/jobs/1'
    });
    const response = generateResponse(request.id);

    jest.spyOn(response, 'json').mockImplementation((data, status) => {
      expect(data).toMatchSnapshot({
        opts: {
          timestamp: expect.any(Number)
        },
        timestamp: expect.any(Number)
      });
      expect(status).toBeUndefined();
      return true;
    });

    return queues[0].add({data: 1}) // Add test job.
      .then(() => jobHandler.getJob(request, response, () => {}, 'test', '1'))
      .then(() => {
        expect(response.json).toBeCalledWith(expect.any(Object));
      });
  });
});
