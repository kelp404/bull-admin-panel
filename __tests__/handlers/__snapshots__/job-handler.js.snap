// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clean jobs that are active 1`] = `"form validation failed."`;

exports[`clean jobs that are delayed 1`] = `"form validation failed."`;

exports[`clean jobs that are paused 1`] = `"form validation failed."`;

exports[`clean jobs that are waiting 1`] = `"form validation failed."`;

exports[`count all state jobs by defaults 1`] = `
Object {
  "active": 0,
  "completed": 0,
  "failed": 0,
  "waiting": 0,
}
`;

exports[`delete job by id 1`] = `Object {}`;

exports[`get job by id 1`] = `
Object {
  "attemptsMade": 0,
  "data": Object {
    "data": 1,
  },
  "delay": 0,
  "failedReason": undefined,
  "finishedOn": undefined,
  "id": "1",
  "name": "__default__",
  "opts": Object {
    "attempts": 1,
    "backoff": undefined,
    "delay": 0,
    "timestamp": Any<Number>,
  },
  "processedOn": undefined,
  "progress": 0,
  "queue": Object {
    "name": "test",
  },
  "returnvalue": null,
  "stacktrace": Array [],
  "state": "waiting",
  "timestamp": Any<Number>,
}
`;

exports[`get jobs by defaults 1`] = `
Object {
  "index": 0,
  "items": Array [
    Object {
      "attemptsMade": 0,
      "data": Object {
        "data": 1,
      },
      "delay": 0,
      "failedReason": undefined,
      "finishedOn": undefined,
      "id": "1",
      "name": "__default__",
      "opts": Object {
        "attempts": 1,
        "backoff": undefined,
        "delay": 0,
        "timestamp": Any<Number>,
      },
      "processedOn": undefined,
      "progress": 0,
      "queue": Object {
        "name": "test",
      },
      "returnvalue": null,
      "stacktrace": Array [],
      "state": undefined,
      "timestamp": Any<Number>,
    },
  ],
  "size": 20,
  "total": 1,
}
`;

exports[`get jobs by failed queries: extra 1`] = `
Array [
  Object {
    "actual": "a",
    "expected": "/^d*$/",
    "field": "index",
    "message": "The 'index' field fails to match the required pattern.",
    "type": "stringPattern",
  },
  Object {
    "actual": "b",
    "expected": "/^d*$/",
    "field": "size",
    "message": "The 'size' field fails to match the required pattern.",
    "type": "stringPattern",
  },
  Object {
    "actual": "c",
    "expected": "waiting, active, completed, failed",
    "field": "state",
    "message": "The 'state' field does not match any of the allowed values.",
    "type": "stringEnum",
  },
]
`;

exports[`get jobs by failed queries: message 1`] = `"form validation failed."`;

exports[`get jobs that are waiting 1`] = `
Object {
  "index": 0,
  "items": Array [
    Object {
      "attemptsMade": 0,
      "data": Object {
        "data": 1,
      },
      "delay": 0,
      "failedReason": undefined,
      "finishedOn": undefined,
      "id": "1",
      "name": "__default__",
      "opts": Object {
        "attempts": 1,
        "backoff": undefined,
        "delay": 0,
        "timestamp": Any<Number>,
      },
      "processedOn": undefined,
      "progress": 0,
      "queue": Object {
        "name": "test",
      },
      "returnvalue": null,
      "stacktrace": Array [],
      "state": undefined,
      "timestamp": Any<Number>,
    },
  ],
  "size": 20,
  "total": 1,
}
`;

exports[`retry job by id 1`] = `Object {}`;
